APM > .NET > add new integrations and update logging (#11623)
DataDog/documentation: content/en/tracing/connect_logs_and_traces/dotnet.md

{{% /tab %}}
{{% tab "Microsoft.Extensions.Logging" %}}
No additional setup is required if you're using _Microsoft.Extensions.Logging_. Trace and span IDs will be automatically injected into application logs using the properties `dd_trace_id` and `dd_span_id`, as long as there is an active trace when the log is written. If there is not an active trace, only `dd_env`, `dd_service`, and `dd_version` properties will be injected.

Note that if you are using a logging framework that replaces the default `LoggerFactory` implementation, such as the [_Serilog.Extensions.Hosting_][1] package, you will need to follow the instructions for that framework (Serilog, in this case).

{{% /tab %}}
{{< /tabs >}}

Next, complete the setup for either automatic or manual injection.

## Automatic injection

The final step to enable automatic correlation identifier injection is:

3. In the .NET Tracer’s environment variables, enable  `DD_LOGS_INJECTION=true`. For alternative ways to configure the .NET Tracer, see [Configuring the .NET Tracer][6].

After you have configured correlation identifier injection, configure log collection as described in [C# Log Collection][7].

## Manual injection

> If you prefer to manually correlate your traces with your logs, manually add correlation identifiers to your logs.

Reviewer: ```suggestion
If you prefer to manually correlate your traces with your logs, you can add correlation identifiers to your logs.
```

[1]: https://github.com/DataDog/dd-trace-dotnet/blob/master/samples/AutomaticTraceIdInjection/NLog40Example/NLog.config
[2]: https://github.com/DataDog/dd-trace-dotnet/blob/master/samples/AutomaticTraceIdInjection/NLog45Example/NLog.config
[3]: https://github.com/DataDog/dd-trace-dotnet/blob/master/samples/AutomaticTraceIdInjection/NLog46Example/NLog.config
{{% /tab %}}
{{% tab "Microsoft.Extensions.Logging" %}}

> If you are using `Microsoft.Extensions.Logging`, no additional setup is required. 

Reviewer 2: Doesn't the customer also have to enable log scopes for this to work? https://docs.microsoft.com/en-us/aspnet/core/fundamentals/logging/?view=aspnetcore-5.0#log-scopes-1

Author: The unfortunate answer is "it depends". It depends entirely on the `ILoggerProvider` implementation. 

The sad reality is there _is_ no file logging provider OOTB (which is why [I wrote one](https://www.nuget.org/packages/NetEscapades.Extensions.Logging.RollingFile) 😄), and as I basically stole the blob storage implementation, mine also requires setting `IncludeScopes=true`.

Which is a long winded way of saying, yes, I think we should add something like:

**Note:** Log injection is only possible with `Microsoft.Extensions.Logging` if your logging provider supports [log scopes](https://docs.microsoft.com/aspnet/core/fundamentals/logging/#log-scopes-1). If log scopes are optional for your provider, be sure to enable them.

Author: Updated to add this in https://github.com/DataDog/documentation/pull/11623/commits/18cb25cb6ac91aedf4fadc23007a199a88beca5b

> 

Reviewer 2: On another note, the ILogger logs injection requires automatic instrumentation. I think we may need to have a separate step for ILogger that ensures that automatic instrumentation is set up. What do you think of changing the getting started section to make the logging library selection the first step and the rest of the content is embedded into each tab (creating some duplication)?

Author: Yes, ILogger requires automatic instrumentation (to be honest, I only recently twigged that the other providers _don't_ require it!) 😄 

It's only a small part of duplication, so I think that's ok. It will also make updating for log4Net easier once we switch that to auto-instrumentation

Author: Made this change in https://github.com/DataDog/documentation/pull/11623/commits/18cb25cb6ac91aedf4fadc23007a199a88beca5b

**Note:** If you are not using a [Datadog Log Integration][7] to parse your logs, custom log parsing rules must parse `dd.trace_id` and `dd.span_id` as strings. For information, see the [FAQ on this topic][8].

> To finish setting up manual log enrichment, after doing the two steps [above](#getting-started):

Reviewer: ```suggestion
After completing Step 1 and 2 [above](#getting-started), finish your manual log enrichment setup:
```

**Note:** If you are not using a [Datadog Log Integration][7] to parse your logs, custom log parsing rules must parse `dd.trace_id` and `dd.span_id` as strings. For information, see the [FAQ on this topic][8].

After completing Step 1 and 2 [above](#getting-started), finish your manual log enrichment setup:

1. Reference the [`Datadog.Trace` NuGet package][9] in your project.

> 2. Use the `CorrelationIdentifier` API to retrieve correlation identifiers and add them to the log context while a span is active.

Reviewer: ```suggestion
2. Use the `CorrelationIdentifier` API to retrieve correlation identifiers and add them to the log context while a span is active.

Lastly, see [C# Log Collection][7] to configure your log collection.
```

}
```

{{% /tab %}}
{{% tab "Microsoft.Extensions.Logging" %}}

```csharp
using Datadog.Trace;
using Microsoft.Extensions.Logging;

ILogger _logger;

// there must be spans started and active before this block.
using(_logger.BeginScope(new Dictionary<string, object>
{
    {"dd.env", CorrelationIdentifier.Env},
    {"dd.service", CorrelationIdentifier.Service},
    {"dd.version", CorrelationIdentifier.Version},
    {"dd.trace_id", CorrelationIdentifier.TraceId.ToString()},
    {"dd.span_id", CorrelationIdentifier.SpanId.ToString()},
}))
{
    // Log something
}
```

{{% /tab %}}
{{< /tabs >}}

> Finally, configure log collection as described in [C# Log Collection][7].

Reviewer: ```suggestion
Lastly, see [C# Log Collection][7] to configure your log collection.
```

Reviewer: @andrewlock because this last instruction appears after the example tables, customers may not scroll down completely to see the last step directing them to C# Log Collection. What are your thoughts on moving this sentence to after Step 4?

Author: I'm fine with that. The downside is the code to do step 4 would no longer directly follow step 4, but I don't think that's a big problem

Reviewer: ```suggestion
```

## Configure log collection

> Ensure log collection is configured in the Datadog Agent, and that the [logs Agent configuration][10] for the specified files to tail, sets `source: csharp`, so that log pipelines can parse the log files . See [C# Log Collection][7] for details.

Reviewer: ```suggestion
Ensure that log collection is configured in the Datadog Agent and that the [Logs Agent configuration][10] for the specified files to tail is set to `source: csharp` so log pipelines can parse the log files. For more information, see [C# Log Collection][7].
```

> <div class="alert alert-info"><strong>Note:</strong> Automatic log collection only works for logs formatted as JSON. Otherwise, use must use custom parsing rules.</div>

Reviewer: ```suggestion
<div class="alert alert-warning"><strong>Note:</strong> Automatic log collection only works for logs formatted as JSON. Alternatively, use custom parsing rules.</div>
```
